www/database.php

```diff
@@ -118,47 +144,70 @@ function modelAction($parameters){
-				$get_q = "select * from session where `group` = '$src' AND problem='$model' AND mode = 'AUTHOR' ORDER BY time DESC limit 1";
-				$res = $this->getDBResults($get_q);
-				if(!$res) {
-					//echo "session selection failed"."<br/>";
+				//step 1: retrieve the most recent session for the specific problem and group which has a corresponding solutions entry
+
+				//$get_sol_q = "select solutions.session_id,solutions.solution_graph from solutions INNER JOIN session ON solutions.session_id = session.session_id AND session.group = '$src' AND session.problem = '$model' AND session.mode = 'AUTHOR' ORDER BY session.time DESC limit 1";
+				$query = $this->getQuery('getSolutionGraph');
+				if($query != ''){
+					$query = sprintf($query, $src, $model, 'AUTHOR');
+				}
+				else{
 					return null;
 				}
-				$data = mysqli_fetch_array($res);
-				$old_session_id = $data["session_id"];
-
-				//step 2: copy solution for the old session id which also has to be a fresh insert
-				$get_sol_q = "select * from solutions where session_id='$old_session_id'";
-				$sol_res = $this->getDBResults($get_sol_q);
+				$sol_res = $this->getDBResults($query);
 				$sol_data = mysqli_fetch_array($sol_res);
-				$solution_graph = $sol_data['solution_graph'];
-				//step 3 : create a new session id with current micro timestamp and session_id
+				$solution_graph = $this->db_connection->real_escape_string($sol_data['solution_graph']);
+				$old_session_id = $sol_data['session_id'];
+
+				//step 2 : create a new session id with current micro timestamp and session_id
 				//This is work around and in future we need a php session creator from username and section like in js
 				$new_sess1 = explode("_",$old_session_id);
 				$milliseconds = round(microtime(true) * 1000);
 				$new_session_id = $new_sess1[0]."_".$milliseconds;
 
-				//step 4 : insert new session
-
+				//step 3 : insert new session
+				/*
 				$create_sess_q = "insert into session(`session_id`,`user`,problem,`mode`,`group`,`section`,`activity`)
 								  VALUES ('$new_session_id','$user','$model','AUTHOR','$dest','$copy_sec','construction')";
-				$create_sess_res = $this->getDBResults($create_sess_q);
-				if(!$create_sess_res)
+				*/
+				$query = $this->getQuery('insertSession');
+				if($query != ''){
+					$query = sprintf($query, $new_session_id, 'AUTHOR', $user, $copy_sec, $model, $dest, 'construction');
+				} else {
 					return null;
+				}
 
+				$create_sess_res = $this->getDBResults($query);
+				if(!$create_sess_res)
+					return null;
+				/*
 				$create_new_sol = "insert into solutions(`session_id`,`share`,`deleted`,`solution_graph`)
 								   VALUES ('$new_session_id',0,0,'$solution_graph')";
-				$create_sol_res = $this->getDBResults($create_new_sol);
+				*/
+				$query = $this->getQuery('insertSolutionGraph');
+				if($query != ''){
+					$query = sprintf($query, $new_session_id, 0, 0, $solution_graph);
+				} else {
+					return null;
```

Comment:
I dont think we need to check this. We are making a local object whose keys we are creating and accessing them. If we return null because of the keys that we created and not due to external data, then we are screwing somewhere, that is we are writing the wrong key. I dont think we need to check it, because it is an object with key value pair that we created. If it is coming as an input from user one might understand this extra case.

Reply:
Do you understand what I am trying to say? Please update it at all the places. If it is wrong in the old code you can udpate it even there.

Reply:
Discussed and decided to remove else cases on slack